File: 1to150/61_rotateList.java
/*
1 2 3 4 5 null, k = 2
假设我们把list连成环状，   -> 1 -> 2 -> 3 -> 4 -> 5 -> 相当于现在的tail.next = 1， 那么如果k = 2， 意味着head，往右移动 n - k = 5 - 2 =3这个地方，所以这意味着，我们找到tail at n - k - 1 = 2这个地方也就是node=3， 将3.next = null, 也就是将链接断开即可。现在的list就变成：
   ->1->2->3- null    4->5->    也就是 4 5 1 2 3 

*/
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null){
            //如果就一个数或者为空，那无论怎么旋转后的结果都是自己
            return head;
        }
        //we need to find how long this linkedlist is, 注意这里是1，因为如果不为空的linkedList，最少长度也是1
        int n = 1;
        ListNode pointer = head;
        /*
        循环找到这个head的tail
        */
        while(pointer.next!= null){
            pointer = pointer.next;
            n++;
        }
        
        //链接tail和head, 此时的pointer已经在尾部了
        pointer.next = head;
        
        /*
        找到新的head， or 其实我们想找新的尾巴 tail @ n - k - 1
        */
        int trueK = k % n;
        for (int i = 0; i < n - trueK - 1; i++){
            head = head.next;
        }
        //现在的head 已经在 newTail这个位置了，所以我们的newHead 指向newTail.next即可
        ListNode newHead = head.next;
        //并且断开newTail后面的pointer
        head.next = null;
        return newHead;
    }
}

File: 1to150/61_rotateList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

'''
1 2 3 4 5 none
5 1 2 3 4 none
4 5 1 2 3 none

1 2 3 | 4 5 1 2 3 4 
如果将首尾相连接后， 在第k位的时候切断link，将3的next指向none即可

newHead在n - k处
所以分割tail在 n - k - 1

但是k需要取余n得到转了多少圈后落到的位置
'''

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        if not head or k == 0:
            return head
        
        dummyTail = head
        dummyHead = head
        
        #找到尾巴
        n = 1
        while(dummyTail.next != None):
            dummyTail = dummyTail.next
            n +=1
            
        #首尾相连
        dummyTail.next = dummyHead
        tail = dummyHead 
        count = 0 
        
        #注意这里我们是要找到分割之后的head在哪里， 那肯定就是 n - k
        #所以我们需要停止在tail，因为只有知道了tail我们才能 新创一个内存head，指向了tail.next ，然后再把tail.next 指向空。
        #所以分割点在n - k - 1
        #但是 k要是大于n了，怎么办，我们需要取余 k % n得到转了多少圈后落到的位置
        
        while(count < n - k % n - 1):
            tail = tail.next
            count += 1
        
        print(tail.val)
        newHead = tail.next
        tail.next = None
        
        return newHead